clients/src/main/java/org/apache/kafka/clients/consumer/StubbedAsyncKafkaConsumer.java

```diff
@@ -0,0 +1,548 @@
+import org.apache.kafka.clients.consumer.internals.events.RequestRebalanceEvent;
+import org.apache.kafka.clients.consumer.internals.events.SubscribePatternEvent;
+import org.apache.kafka.clients.consumer.internals.events.SubscribeTopicsEvent;
+import org.apache.kafka.clients.consumer.internals.events.UnsubscribeEvent;
+import org.apache.kafka.common.Metric;
+import org.apache.kafka.common.MetricName;
+import org.apache.kafka.common.PartitionInfo;
+import org.apache.kafka.common.TopicPartition;
+import org.apache.kafka.common.errors.InvalidGroupIdException;
+import org.apache.kafka.common.errors.RecordDeserializationException;
+import org.apache.kafka.common.header.Headers;
+import org.apache.kafka.common.header.internals.RecordHeaders;
+import org.apache.kafka.common.record.Record;
+import org.apache.kafka.common.record.TimestampType;
+import org.apache.kafka.common.requests.ListOffsetsRequest;
+import org.apache.kafka.common.serialization.Deserializer;
+import org.apache.kafka.common.utils.Time;
+import org.apache.kafka.common.utils.Timer;
+import org.apache.kafka.common.utils.Utils;
+
+import java.nio.ByteBuffer;
+import java.time.Duration;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.OptionalLong;
+import java.util.Set;
+import java.util.regex.Pattern;
+
+/**
+ * This {@link StubbedAsyncKafkaConsumer stub} implementation of {@link Consumer} serves as a <i>wireframe</i> or
+ * <i>sketch</i>, showing the interaction between the foreground and background threads. Each of the main API methods
+ * will need to answer the following questions:
+ *
+ * <ol>
+ *     <li>Does this method block?</li>
+ *     <li>Does this method interact with the background thread?</li>
+ *     <li>If yes, what data is passed as input to the background thread?</li>
+ *     <li>If yes, what data is returned as output from the background thread?</li>
+ * </ol>
+ *
+ * @param <K> Key
+ * @param <V> Value
+ * @see ApplicationEventProcessor for the logic of the background event handler
+ */
+public class StubbedAsyncKafkaConsumer<K, V> implements Consumer<K, V> {
+
+    /**
+     * These instance variables are intentionally left unassigned, to avoid clutter...
+     */
+    private Time time;
+
+    private EventHandler eventHandler;
+
+    private SubscriptionState subscriptions;
+
+    private Deserializer<K> keyDeserializer;
+
+    private Deserializer<V> valueDeserializer;
+
+    private long defaultApiTimeoutMs;
+
+    private List<ConsumerPartitionAssignor> assignors;
+
+    private Optional<String> groupId;
+
+    /**
+     * Answers to the above questions:
+     *
+     * <ol>
+     *     <li>No</li>
+     *     <li>No</li>
+     *     <li><i>n/a</i></li>
+     *     <li><i>n/a</i></li>
+     * </ol>
+     */
+    @Override
+    public Set<TopicPartition> assignment() {
```

Comment:
While I agree this func should be non-blocking, I still want to clarify the ordering restrictions of this func when used together with others like `poll`. For example, assume the following events happened at the background thread side with `subscribe` scenarios:

1) Currently assigned with partition A and B.
2) Rebalance happened, and the new assignment updated to C and D.
3) Sent a notification via the foreground queue about the new assignment, to let the foreground thread trigger the corresponding rebalance listener.
4) Dropped all records in the raw records buffer for partition A and B, fetched new data from partition C and D.


Now suppose the foreground thread calls `assignment` between time 2) and 3), it would still return `{A, B}`, and then later when it calls `poll` between time 3) and 4), it should immediately trigger the rebalance listener, and then records from C and D can be returned later.

If the foreground thread calls `assignment` between time 3) and 4), the foreground thread already knows about the new assignment so it could return `{C, D}`. HOWEVER, since it has not yet calls `poll` it would not trigger the rebalance listener --- this is because in docs we state that

```
This callback will only execute in the user thread as part of the {@link Consumer#poll(java.time.Duration) poll(long)} call whenever partition assignment changes.
```

Which is a bit weird since it is out-of-ordering from the caller's perspective: when I call `assignment()` and then `poll()`, I would first already get the new assignment, but the listener indicating that assignment as changed via `onPartitionsRevoked/Assigned` would only come later in the `poll`.

While searching on the doc I did not find anywhere that we state this ordering would be guaranteed, I concern there are users who's assuming this ordering in their app development.

Reply:
If we want to still respect the ordering of assignment changes from caller's perspective here, I think we'd consider 1) let the `assignment()` call still return the stale assignment before the rebalance listener is triggered, to be consistent with the current behavior; 2) change the `listener` API that this may also be triggered in other funcs than `poll`, e.g. in `assignment` as well.

Personally I'd prefer we do 1), but would love to hear your thoughts as well @hachikuji @philipnee

Reply:
I feel we need an explicit synchronization barrier to ensure state changes happen in a deterministic sequence.  I mentioned this idea in our previous meeting, but I just want to bring it up again for discussion: (feel free to strike down this idea)

Is it reasonable to ask users to invoke poll() after any sort of assignment to trigger the rebalance? I think we could use poll() as an explicit barrier to synchronize both threads.

So back to your example,
1. User updates assignment via `assign()`
2. Read calls (`assignments()`) here will still returns the previous state
3.  client invoke poll() - update client and background state and triggers rebalance. Rebalance callback might be triggered here.
4. `assignments()` returns the updated version of the assignment. Invalid fetches will be dropped here.

I think it kind of breaks the current contract, as `assign()` -> `assignment()` should already reflect to the current assignment.  But I do think regardless of how we handle this, there will be some kind of behavioral changes as the threading model is now asynchronous, which kind of implies data update is expected to be asynchronous.

Reply:
Afterall, I think it is a worse user experience if the assignment call cannot yield deterministic results. Even though we will need a KIP for this behavior change, I think it kind of provides the guarantee to the user.

Reply:
In the `assign()` -> `assignment()` example above, there is no "rebalance" triggered and hence there's no rebalance listener triggered either. The rebalance listener is only relevant with the `subscribe()` scenarios.

So just to summarize all the scenarios here to avoid confusions between us :)

1. `assign()`: the followed `assignment()`  should always reflect the latest partition list from the previous `assign()`, and also the followed `poll()` should return data from the latest partition list only (i.e. the background thread, upon getting this new partition list, should cleanup the buffered records accordingly, and the followed `poll` call should return empty data if no new data is retrieved within the poll timeout yet).
2. `subscribe()`: the followed `assignment()` could potentially still return current assignment which is inconsistent with the `subscribe()` --- this is the case today, but as I mentioned above I think could be open for discussion whether we can to change  --- or return the new assignment after the rebalance if enough time has elapsed so that the rebalance is done; the followed `poll()` though should also only return data from the latest partition list as a result of the new subscription.
3. `seek/seekToB/E`: the followed `position` should reflect the newly set position, and also the followed `poll` should only return data from the new seeking position.

Reply:
Oh my mistake, entirely forgot about there's no rebalancing after the assignment().